src/neighborsearch/filtered.jl

```diff
@@ -0,0 +1,193 @@
+    # nmax and maxpersector checks
+    if categs[:use]
+      cat, pass = Dict(), false
+      tab = Tables.columns(values(obj))
+      for col in keys(categs[:max])
+        cat[col] = Tables.getcolumn(tab, col)[i]
+        categs[:count][col][cat[col]] >= categs[:max][col] && (pass = true)
+      end
+      pass && continue
+    end
+
+    # check sectors; if there are no space for neighbors in the indsector, it's
+    # ignored; otherwise sectors[:count][indsector] is incremented in the end
+    # after nmax checks
+    if sectors[:use]
+      centered  = sectors[:rotmat]' * (centroid(obj,i) - pₒ)
+      indsector = getsector(centered)
+      sectors[:count][indsector] >= sectors[:max] && continue
+    end
+
+    # add neighbor
+    nneigh += 1
+    if nmax == Inf
+      push!(neighbors, i)
+    else
+      neighbors[nneigh] = i
+    end
+
+    nneigh == nmax && break
+
+    # add counters
+    sectors[:use] && (sectors[:count][indsector] += 1)
+    if categs[:use]
+      for col in keys(categs[:max])
+        categs[:count][col][cat[col]] += 1
+      end
+    end
+  end
+
+  # slice neighbors if nmax was not reached
+  length(neighbors) > nneigh && (neighbors = neighbors[1:nneigh])
+
+  nneigh
+end
+
+# initialize categories constraints if necessary. e.g. if `maxpercategory =
+# (holeid = 2,)`, dict[:count][:holeid] will have all the possible values of
+# holeid, initiliazed to zero for further counting.
+function initcategories(obj, inds, catgs)
+  catgs == nothing && return Dict(:use => false)
+  tab   = Tables.columns(values(obj))
+  catvals = Dict(k => unique(Tables.getcolumn(tab, k)[inds]) for k in keys(catgs))
+  counter = Dict(k => Dict(zip(v, zeros(Int,length(v)))) for (k, v) in catvals)
+  Dict(:use => true, :max => catgs, :count => counter)
+end
+
+# initialize sectors constraints if necessary. this will create a counter for
+# each sector and a rotation matrix to reverse ellipsoid rotation (if an
+# ellipsoid search is used). a rotation noise is also added to simplify filtering.
+# it is used here because if data is regular and fall onto axes (e.g. cartesian
+# grid data), the sector filtering will return misleading results without it.
+function initsectors(method, maxpersector)
+  maxpersector == nothing && return Dict(:use => false)
+  N = embeddim(method.domain)
+  ellips = method isa NeighborhoodSearch && method.neigh isa Ellipsoid
+
+  # add a noise rotation to avoid samples at sectors limits
+  rotangs = N == 2 ? [0.001] : [0.001, 0.001, 0.001]
+  convention = ellips ? method.neigh.convention : Datamine
+
+  # reverse rotation if it is an ellipsoid
+  if ellips
+    extrinsic = isextrinsic(method.neigh.convention)
+    invertfac = extrinsic ? -1 : 1
+    rotangs   = invertfac * (method.neigh.angles .+ rotangs)
+  end
+
+  # rotation matrix
+  rotmatx = rotmat(rotangs, convention)'
+
+  Dict(:use => true, :count => zeros(Int, 2^N), :rotmat => rotmatx, :max => maxpersector)
+end
```

Comment:
I think the same logic I wrote down for `initcategories` could be applied here. We usually don't need to introduce dictionaries for filtering. We can save memory allocations by writing for loops manually that are well-documented so that people can follow the logic easily. The dictionaries are hiding the simple logic explained in the docstring.